File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        
        if(!root || root == p || root == q) return root;

        TreeNode* left = lowestCommonAncestor(root->left,p,q);
        TreeNode* right = lowestCommonAncestor(root->right,p,q);
        if (left&&right) return root;
        return left? left: right;

        /*
        while (root)
        {
            if (p->val < root->val &&
                q->val < root->val)
            {
                root = root->left;
            }
            else if (p->val > root->val &&
                     q->val > root->val)
            {
                root = root->right;
            }
            else
            {
                return root;
            }
        }
            return nullptr;
        */

    }
};

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 
public class Solution {
    public List<List<int>> LevelOrder(TreeNode root) {
        List<List<int>> result = new();
        Queue<TreeNode> queue = new();
        queue.Enqueue(root);
        while(queue.Count > 0)
        {
            int nodeNumOfLevel = queue.Count;
            List<int> nodes = new ();
            for(int i =0; i<nodeNumOfLevel;i++)
            {
                TreeNode node = queue.Dequeue();
                if(node == null) continue;
                nodes.Add(node.val);
                queue.Enqueue(node.left);
                queue.Enqueue(node.right);
            }
            if(nodes.Count > 0) result.Add(nodes);
        }
        return result;
    }
}
